scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt

```diff
@@ -0,0 +1,209 @@
+ */
+class ClearlyDefinedStorage(
+    /** The configuration for this storage implementation. */
+    val configuration: ClearlyDefinedStorageConfiguration
+) : ScanResultsStorage() {
+    /** The service for interacting with ClearlyDefined. */
+    private val clearlyDefinedService = ClearlyDefinedService.create(configuration.serverUrl)
+
+    override fun readFromStorage(id: Identifier): Result<ScanResultContainer> =
+        readPackageFromClearlyDefined(id, null, null)
+
+    override fun readFromStorage(pkg: Package, scannerDetails: ScannerDetails): Result<ScanResultContainer> =
+        readPackageFromClearlyDefined(pkg.id, pkg.vcs, pkg.sourceArtifact.takeIf { it.url.isNotEmpty() })
+
+    override fun addToStorage(id: Identifier, scanResult: ScanResult): Result<Unit> =
+        Failure("Storing results in ClearlyDefined is not supported.")
+
+    /**
+     * Try to obtain a [ScanResult] produced by ScanCode from ClearlyDefined for the package with the given [id].
+     * Use the VCS information [vcs] and an optional [sourceArtifact] to determine the coordinates for looking up
+     * the result.
+     */
+    private fun readPackageFromClearlyDefined(
+        id: Identifier,
+        vcs: VcsInfo?,
+        sourceArtifact: RemoteArtifact?
+    ): Result<ScanResultContainer> =
+        try {
+            readFromClearlyDefined(id, packageCoordinates(id, vcs, sourceArtifact))
+        } catch (e: IllegalArgumentException) {
+            log.warn("Could not obtain ClearlyDefined coordinates for package $id", e)
+            emptyResult(id)
+        }
+
+    /**
+     * Try to obtain a [ScanResult] produced by ScanCode from ClearlyDefined for the package with the given [id],
+     * using the given [coordinates] for looking up the data. These may not necessarily be equivalent to the
+     * identifier, as we try to lookup source code results if a GitHub repository is known.
+     */
+    private fun readFromClearlyDefined(
+        id: Identifier,
+        coordinates: ClearlyDefinedService.Coordinates
+    ): Result<ScanResultContainer> {
+        val startTime = Instant.now()
+        log.info("Looking up results for ${id.toCoordinates()}.")
+
+        return try {
+            val packageUrl = coordinates.toString()
+            val tools = execute(
+                clearlyDefinedService.harvestTools(
+                    coordinates.type,
+                    coordinates.provider,
+                    coordinates.namespace.orEmpty(),
+                    coordinates.name,
+                    coordinates.revision.orEmpty()
+                )
+            )
+            val scanCodeVersion = findScanCodeVersion(tools, packageUrl)
+            scanCodeVersion?.let { version ->
+                loadScanCodeResults(id, coordinates, version, startTime)
+            } ?: emptyResult(id)
+        } catch (e: IOException) {
+            log.error("Error when reading results for package $id from ClearlyDefined", e)
```

Comment:
Please use `'${id.toCoordinates()}'` when printing `Identifier`s. Also: Missing dot at end of sentence 😀